File: swift-bridge/Sources/CoreDataBridge/FetchedResultsController.swift
import CoreData
import Foundation

@_cdecl("cd_fetched_results_controller_new")
public func cdFetchedResultsControllerNew(
    _ fetchRequestPtr: UnsafeMutableRawPointer?,
    _ managedObjectContextPtr: UnsafeMutableRawPointer?,
    _ sectionNameKeyPath: UnsafePointer<CChar>?,
    _ cacheName: UnsafePointer<CChar>?,
    _ outController: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let fetchRequestPtr, let managedObjectContextPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing fetched-results fetch request or managed object context")
        }
        let fetchRequest: NSFetchRequest<NSManagedObject> = cdBorrow(fetchRequestPtr)
        let managedObjectContext: NSManagedObjectContext = cdBorrow(managedObjectContextPtr)
        let controller = NSFetchedResultsController(
            fetchRequest: fetchRequest,
            managedObjectContext: managedObjectContext,
            sectionNameKeyPath: cdOptionalString(sectionNameKeyPath),
            cacheName: cdOptionalString(cacheName)
        )
        outController?.pointee = cdRetain(controller)
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_fetched_results_controller_perform_fetch")
public func cdFetchedResultsControllerPerformFetch(
    _ controllerPtr: UnsafeMutableRawPointer?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let controllerPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing fetched results controller")
        }
        let controller: NSFetchedResultsController<NSManagedObject> = cdBorrow(controllerPtr)
        try controller.performFetch()
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_fetched_results_controller_get_fetch_request")
public func cdFetchedResultsControllerGetFetchRequest(_ controllerPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let controllerPtr else {
        return nil
    }
    let controller: NSFetchedResultsController<NSManagedObject> = cdBorrow(controllerPtr)
    return cdRetain(controller.fetchRequest)
}

@_cdecl("cd_fetched_results_controller_get_managed_object_context")
public func cdFetchedResultsControllerGetManagedObjectContext(_ controllerPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let controllerPtr else {
        return nil
    }
    let controller: NSFetchedResultsController<NSManagedObject> = cdBorrow(controllerPtr)
    return cdRetain(controller.managedObjectContext)
}

@_cdecl("cd_fetched_results_controller_get_section_name_key_path")
public func cdFetchedResultsControllerGetSectionNameKeyPath(_ controllerPtr: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let controllerPtr else {
        return nil
    }
    let controller: NSFetchedResultsController<NSManagedObject> = cdBorrow(controllerPtr)
    guard let value = controller.sectionNameKeyPath else {
        return nil
    }
    return cdCString(value)
}

@_cdecl("cd_fetched_results_controller_get_cache_name")
public func cdFetchedResultsControllerGetCacheName(_ controllerPtr: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let controllerPtr else {
        return nil
    }
    let controller: NSFetchedResultsController<NSManagedObject> = cdBorrow(controllerPtr)
    guard let value = controller.cacheName else {
        return nil
    }
    return cdCString(value)
}

@_cdecl("cd_fetched_results_controller_get_fetched_objects")
public func cdFetchedResultsControllerGetFetchedObjects(_ controllerPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let controllerPtr else {
        return nil
    }
    let controller: NSFetchedResultsController<NSManagedObject> = cdBorrow(controllerPtr)
    guard let fetchedObjects = controller.fetchedObjects else {
        return nil
    }
    return cdRetain(fetchedObjects as NSArray)
}

@_cdecl("cd_fetched_results_controller_get_sections")
public func cdFetchedResultsControllerGetSections(_ controllerPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let controllerPtr else {
        return nil
    }
    let controller: NSFetchedResultsController<NSManagedObject> = cdBorrow(controllerPtr)
    guard let sections = controller.sections else {
        return nil
    }
    return cdRetain(sections as NSArray)
}

@_cdecl("cd_fetched_results_controller_get_section_index_titles_json")
public func cdFetchedResultsControllerGetSectionIndexTitlesJSON(
    _ controllerPtr: UnsafeMutableRawPointer?,
    _ outJSON: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let controllerPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing fetched results controller")
        }
        let controller: NSFetchedResultsController<NSManagedObject> = cdBorrow(controllerPtr)
        outJSON?.pointee = cdCString(try cdEncodeJSON(controller.sectionIndexTitles))
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_fetched_results_controller_object_at_index_path")
public func cdFetchedResultsControllerObjectAtIndexPath(
    _ controllerPtr: UnsafeMutableRawPointer?,
    _ section: Int64,
    _ item: Int64,
    _ outObject: UnsafeMutablePointer<UnsafeMutableRawPointer?>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let controllerPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing fetched results controller")
        }
        let controller: NSFetchedResultsController<NSManagedObject> = cdBorrow(controllerPtr)
        guard section >= 0, item >= 0 else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Negative fetched results index path")
        }
        let sectionIndex = Int(section)
        let itemIndex = Int(item)
        guard let sections = controller.sections, sectionIndex < sections.count else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Fetched results section out of range")
        }
        guard itemIndex < sections[sectionIndex].numberOfObjects else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Fetched results item out of range")
        }
        let object = controller.object(at: IndexPath(indexes: [sectionIndex, itemIndex]))
        outObject?.pointee = cdRetain(object)
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_fetched_results_controller_index_path_for_object")
public func cdFetchedResultsControllerIndexPathForObject(
    _ controllerPtr: UnsafeMutableRawPointer?,
    _ objectPtr: UnsafeMutableRawPointer?,
    _ outSection: UnsafeMutablePointer<Int64>?,
    _ outItem: UnsafeMutablePointer<Int64>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let controllerPtr, let objectPtr else {
            throw cdBridgeNSError(code: CDR_INVALID_ARGUMENT, message: "Missing fetched results controller or object")
        }
        let controller: NSFetchedResultsController<NSManagedObject> = cdBorrow(controllerPtr)
        let object: NSManagedObject = cdBorrow(objectPtr)
        let indexPath = controller.indexPath(forObject: object)
        if let indexPath, indexPath.count >= 2 {
            outSection?.pointee = Int64(indexPath[0])
            outItem?.pointee = Int64(indexPath[1])
        } else {
            outSection?.pointee = -1
            outItem?.pointee = -1
        }
        return CDR_OK
    } catch let error as NSError {
        cdWriteError(error, to: outError)
        return Int32(error.code)
    }
}

@_cdecl("cd_fetched_results_controller_delete_cache_with_name")
public func cdFetchedResultsControllerDeleteCacheWithName(_ name: UnsafePointer<CChar>?) {
    NSFetchedResultsController<NSManagedObject>.deleteCache(withName: cdOptionalString(name))
}

private func cdFetchedResultsSectionInfo(_ sectionInfoPtr: UnsafeMutableRawPointer) -> any NSFetchedResultsSectionInfo {
    guard let sectionInfo = cdBorrowBox(sectionInfoPtr).object as? (NSObject & NSFetchedResultsSectionInfo) else {
        fatalError("Unexpected fetched results section info type")
    }
    return sectionInfo
}

@_cdecl("cd_fetched_results_section_info_get_name")
public func cdFetchedResultsSectionInfoGetName(_ sectionInfoPtr: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let sectionInfoPtr else {
        return nil
    }
    return cdCString(cdFetchedResultsSectionInfo(sectionInfoPtr).name)
}

@_cdecl("cd_fetched_results_section_info_get_index_title")
public func cdFetchedResultsSectionInfoGetIndexTitle(_ sectionInfoPtr: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    guard let sectionInfoPtr else {
        return nil
    }
    guard let value = cdFetchedResultsSectionInfo(sectionInfoPtr).indexTitle else {
        return nil
    }
    return cdCString(value)
}

@_cdecl("cd_fetched_results_section_info_get_number_of_objects")
public func cdFetchedResultsSectionInfoGetNumberOfObjects(_ sectionInfoPtr: UnsafeMutableRawPointer?) -> UInt64 {
    guard let sectionInfoPtr else {
        return 0
    }
    return UInt64(cdFetchedResultsSectionInfo(sectionInfoPtr).numberOfObjects)
}

@_cdecl("cd_fetched_results_section_info_get_objects")
public func cdFetchedResultsSectionInfoGetObjects(_ sectionInfoPtr: UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? {
    guard let sectionInfoPtr else {
        return nil
    }
    guard let objects = cdFetchedResultsSectionInfo(sectionInfoPtr).objects else {
        return nil
    }
    return cdRetain(objects as NSArray)
}

File: src/fetched_results_controller.rs
use crate::context::{NSManagedObject, NSManagedObjectContext};
use crate::error::CoreDataError;
use crate::ffi;
use crate::private::{
    collect_array, cstring_from_str, error_from_status, impl_object_wrapper, parse_json_ptr,
    take_string,
};
use crate::query::NSFetchRequest;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[non_exhaustive]
pub enum NSFetchedResultsChangeType {
    Insert,
    Delete,
    Move,
    Update,
    Unknown(u64),
}

impl NSFetchedResultsChangeType {
    pub const fn from_raw(raw: u64) -> Self {
        match raw {
            1 => Self::Insert,
            2 => Self::Delete,
            3 => Self::Move,
            4 => Self::Update,
            other => Self::Unknown(other),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct FetchedResultsIndexPath {
    pub section: usize,
    pub item: usize,
}

impl FetchedResultsIndexPath {
    pub const fn new(section: usize, item: usize) -> Self {
        Self { section, item }
    }
}

impl_object_wrapper!(NSFetchedResultsController);
impl_object_wrapper!(NSFetchedResultsSectionInfo);

impl NSFetchedResultsController {
    pub fn new(
        fetch_request: &NSFetchRequest,
        managed_object_context: &NSManagedObjectContext,
        section_name_key_path: Option<&str>,
        cache_name: Option<&str>,
    ) -> Result<Self, CoreDataError> {
        let section_name_key_path = section_name_key_path
            .map(|value| cstring_from_str(value, "fetched results section name key path"))
            .transpose()?;
        let cache_name = cache_name
            .map(|value| cstring_from_str(value, "fetched results cache name"))
            .transpose()?;
        let mut out_controller = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_fetched_results_controller_new(
                fetch_request.as_ptr(),
                managed_object_context.as_ptr(),
                section_name_key_path
                    .as_ref()
                    .map_or(core::ptr::null(), |value| value.as_c_str().as_ptr()),
                cache_name
                    .as_ref()
                    .map_or(core::ptr::null(), |value| value.as_c_str().as_ptr()),
                &mut out_controller,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { Self::from_retained_ptr(out_controller, "fetched results controller") }
    }

    pub fn perform_fetch(&self) -> Result<(), CoreDataError> {
        let mut out_error = core::ptr::null_mut();
        let status = unsafe { ffi::cd_fetched_results_controller_perform_fetch(self.as_ptr(), &mut out_error) };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        Ok(())
    }

    pub fn fetch_request(&self) -> Result<NSFetchRequest, CoreDataError> {
        let ptr = unsafe { ffi::cd_fetched_results_controller_get_fetch_request(self.as_ptr()) };
        unsafe { NSFetchRequest::from_retained_ptr(ptr, "fetched results controller fetch request") }
    }

    pub fn managed_object_context(&self) -> Result<NSManagedObjectContext, CoreDataError> {
        let ptr =
            unsafe { ffi::cd_fetched_results_controller_get_managed_object_context(self.as_ptr()) };
        unsafe {
            NSManagedObjectContext::from_retained_ptr(
                ptr,
                "fetched results controller managed object context",
            )
        }
    }

    pub fn section_name_key_path(&self) -> Option<String> {
        let ptr = unsafe { ffi::cd_fetched_results_controller_get_section_name_key_path(self.as_ptr()) };
        unsafe { take_string(ptr) }
    }

    pub fn cache_name(&self) -> Option<String> {
        let ptr = unsafe { ffi::cd_fetched_results_controller_get_cache_name(self.as_ptr()) };
        unsafe { take_string(ptr) }
    }

    pub fn fetched_objects(&self) -> Result<Vec<NSManagedObject>, CoreDataError> {
        let array_ptr = unsafe { ffi::cd_fetched_results_controller_get_fetched_objects(self.as_ptr()) };
        collect_array(array_ptr, "fetched results controller objects")
    }

    pub fn sections(&self) -> Result<Vec<NSFetchedResultsSectionInfo>, CoreDataError> {
        let array_ptr = unsafe { ffi::cd_fetched_results_controller_get_sections(self.as_ptr()) };
        collect_array(array_ptr, "fetched results controller sections")
    }

    pub fn section_index_titles(&self) -> Result<Vec<String>, CoreDataError> {
        let mut out_json = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_fetched_results_controller_get_section_index_titles_json(
                self.as_ptr(),
                &mut out_json,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { parse_json_ptr(out_json, "fetched results controller section index titles") }
    }

    pub fn object_at_index_path(
        &self,
        index_path: FetchedResultsIndexPath,
    ) -> Result<NSManagedObject, CoreDataError> {
        let section = i64::try_from(index_path.section)
            .map_err(|_| CoreDataError::bridge(-1, "fetched results section overflow"))?;
        let item = i64::try_from(index_path.item)
            .map_err(|_| CoreDataError::bridge(-1, "fetched results item overflow"))?;
        let mut out_object = core::ptr::null_mut();
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_fetched_results_controller_object_at_index_path(
                self.as_ptr(),
                section,
                item,
                &mut out_object,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        unsafe { NSManagedObject::from_retained_ptr(out_object, "fetched results controller object") }
    }

    pub fn index_path_for_object(
        &self,
        object: &NSManagedObject,
    ) -> Result<Option<FetchedResultsIndexPath>, CoreDataError> {
        let mut section = -1_i64;
        let mut item = -1_i64;
        let mut out_error = core::ptr::null_mut();
        let status = unsafe {
            ffi::cd_fetched_results_controller_index_path_for_object(
                self.as_ptr(),
                object.as_ptr(),
                &mut section,
                &mut item,
                &mut out_error,
            )
        };
        if status != ffi::status::OK {
            return Err(unsafe { error_from_status(status, out_error) });
        }
        if section < 0 || item < 0 {
            return Ok(None);
        }
        let section = usize::try_from(section)
            .map_err(|_| CoreDataError::bridge(-1, "negative fetched results section"))?;
        let item = usize::try_from(item)
            .map_err(|_| CoreDataError::bridge(-1, "negative fetched results item"))?;
        Ok(Some(FetchedResultsIndexPath::new(section, item)))
    }

    pub fn delete_cache_with_name(name: Option<&str>) -> Result<(), CoreDataError> {
        let name = name
            .map(|value| cstring_from_str(value, "fetched results cache name"))
            .transpose()?;
        unsafe {
            ffi::cd_fetched_results_controller_delete_cache_with_name(
                name.as_ref()
                    .map_or(core::ptr::null(), |value| value.as_c_str().as_ptr()),
            );
        }
        Ok(())
    }
}

impl NSFetchedResultsSectionInfo {
    pub fn name(&self) -> Result<String, CoreDataError> {
        let ptr = unsafe { ffi::cd_fetched_results_section_info_get_name(self.as_ptr()) };
        unsafe { take_string(ptr) }
            .ok_or_else(|| CoreDataError::bridge(-1, "fetched results section name was nil"))
    }

    pub fn index_title(&self) -> Option<String> {
        let ptr = unsafe { ffi::cd_fetched_results_section_info_get_index_title(self.as_ptr()) };
        unsafe { take_string(ptr) }
    }

    pub fn number_of_objects(&self) -> usize {
        unsafe { ffi::cd_fetched_results_section_info_get_number_of_objects(self.as_ptr()) as usize }
    }

    pub fn objects(&self) -> Result<Vec<NSManagedObject>, CoreDataError> {
        let array_ptr = unsafe { ffi::cd_fetched_results_section_info_get_objects(self.as_ptr()) };
        collect_array(array_ptr, "fetched results section objects")
    }
}
